HW3/P3/sum.cl

```diff
@@ -24,11 +27,24 @@ __kernel void sum_coalesced(__global float* x,
     // You can assume get_local_size(0) is a power of 2.
     //
     // See http://www.nehalemlabs.net/prototype/blog/2014/06/16/parallel-programming-with-opencl-and-python-parallel-reduce/
-    for (;;) { // YOUR CODE HERE
-        ; // YOUR CODE HERE
-    }
 
+    //
+
+    int offset = get_local_size(0);
+
+     while (offset != 1)
+    {
+        offset = offset >> 1;
+        fast[local_id] = fast[local_id] + fast[local_id+offset];
```

Comment:
You are missing an if statement before this line.
You should only increment fast[local_id] if local_id < offset, otherwise, unpredicted result may occur (reading from fast[local_id+offset] that is being updated by the thread at index local_id+offset.